File: core/aws_ddk_core/stages/sqs_lambda.py
# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import List, Optional

from aws_cdk import Duration
from aws_cdk.aws_events import EventPattern, IRuleTarget
from aws_cdk.aws_events_targets import SqsQueue
from aws_cdk.aws_iam import Effect, IRole, PolicyStatement
from aws_cdk.aws_lambda import Code, IFunction, ILayerVersion, Runtime
from aws_cdk.aws_lambda_event_sources import SqsEventSource
from aws_cdk.aws_sqs import DeadLetterQueue, IQueue
from aws_ddk_core.pipelines.stage import DataStage
from aws_ddk_core.resources import LambdaFactory, SQSFactory
from constructs import Construct


class SqsToLambdaStage(DataStage):
    """
    Class that represents an SQS to Lambda Transform DDK Stage.
    """

    def __init__(
        self,
        scope: Construct,
        id: str,
        environment_id: str,
        code: Optional[Code] = None,
        handler: Optional[str] = None,
        runtime: Runtime = Runtime.PYTHON_3_9,
        role: Optional[IRole] = None,
        memory_size: Optional[int] = None,
        timeout: Optional[Duration] = None,
        visibility_timeout: Optional[Duration] = None,
        dead_letter_queue_enabled: bool = False,
        max_receive_count: int = 1,
        batch_size: Optional[int] = None,
        layers: Optional[List[ILayerVersion]] = None,
        lambda_function: Optional[IFunction] = None,
        sqs_queue: Optional[IQueue] = None,
        lambda_function_errors_alarm_threshold: Optional[int] = 5,
        lambda_function_errors_alarm_evaluation_periods: Optional[int] = 1,
    ) -> None:
        """
        DDK SQS to Lambda stage.

        It implements an Amazon SQS queue connected to an AWS Lambda function, with an optional DLQ.

        Parameters
        ----------
        scope : Construct
            Scope within which this construct is defined
        id : str
            Identifier of the stage
        environment_id : str
            Identifier of the environment
        code : Optional[Code]
            The source code of the Lambda function
            Must be set if `lambda_function` is not.
        handler : Optional[str]
            The name of the method within the code that Lambda calls to execute the function.
            Must be set if `lambda_function` is not.
        runtime : Runtime
            The runtime environment for the Lambda function. `PYTHON_3_9` by default
        role : Optional[IRole]
            Lambda execution role
        memory_size : Optional[int]
            The amount of memory, in MB, that is allocated to the Lambda function. `256` by default
        timeout : Optional[Duration]
            The function execution time (in seconds) after which Lambda terminates the function.
            `aws_cdk.Duration.seconds(120)` by default
        visibility_timeout : Optional[Duration]
            Timeout of processing a single message in the queue. Default: Duration.seconds(120).
        dead_letter_queue_enabled : bool
            Determines if DLQ is enabled. `False` by default
        max_receive_count : int
            The number of times a message can be unsuccessfully dequeued before being moved to the dead-letter queue.
            `1` by default
        batch_size : Optional[int]
            The maximum number of records retrieved from the event source at the function invocation time.
            `10` by default
        layers : Optional[List[ILayerVersion]]
            A list of layers to add to the lambda function's execution environment.
        lambda_function: Optional[IFunction]
            Preexisting Lambda Function to use in stage. `None` by default
        sqs_queue: Optional[IQueue]
            Preexisting SQS Queue  to use in stage. `None` by default
        lambda_function_errors_alarm_threshold: Optional[int]
            Amount of errored function invocations before triggering CW alarm. Defaults to `5`
        lambda_function_errors_alarm_evaluation_periods: Optional[int]
            The number of periods over which data is compared to the specified threshold. Defaults to `1`
        """
        super().__init__(scope, id)

        self._event_source: str = f"{id}-event-source"
        self._event_detail_type: str = f"{id}-event-type"

        if lambda_function:
            self._function = lambda_function
        elif code and handler:
            self._function = LambdaFactory.function(
                self,
                id=f"{id}-function",
                environment_id=environment_id,
                code=code,
                handler=handler,
                runtime=runtime,
                role=role,
                memory_size=memory_size,
                timeout=timeout,
                environment={
                    "EVENT_SOURCE": self._event_source,
                    "EVENT_DETAIL_TYPE": self._event_detail_type,
                },
                layers=layers,
            )
        else:
            raise ValueError("'code' and 'handler' or 'lambda_function' must be set to instantiate this stage")

        # Enable the function to publish events to the default EventBus
        self._function.add_to_role_policy(
            PolicyStatement(
                effect=Effect.ALLOW,
                actions=[
                    "events:PutEvents",
                ],
                resources=["*"],
            )
        )

        self._dlq: Optional[DeadLetterQueue] = None
        if dead_letter_queue_enabled:
            self._dlq = DeadLetterQueue(
                max_receive_count=max_receive_count,
                queue=SQSFactory.queue(
                    self,
                    id=f"{id}-dlq",
                    environment_id=environment_id,
                ),
            )

        self._queue = sqs_queue or SQSFactory.queue(
            self,
            id=f"{id}-queue",
            environment_id=environment_id,
            visibility_timeout=visibility_timeout,
            dead_letter_queue=self._dlq,
        )

        self._function.add_event_source(SqsEventSource(queue=self._queue, batch_size=batch_size))

        self.add_alarm(
            alarm_id=f"{id}-function-errors",
            alarm_metric=self._function.metric_errors(),
            alarm_threshold=lambda_function_errors_alarm_threshold,
            alarm_evaluation_periods=lambda_function_errors_alarm_evaluation_periods,
        )

    @property
    def function(self) -> IFunction:
        """
        Return: Function
            The Lambda function
        """
        return self._function

    @property
    def queue(self) -> IQueue:
        """
        Return: Queue
            The SQS queue
        """
        return self._queue

    @property
    def dlq(self) -> Optional[DeadLetterQueue]:
        """
        Return: DeadLetterQueue
            The SQS dead letter queue
        """
        return self._dlq

    def get_event_pattern(self) -> Optional[EventPattern]:
        return EventPattern(
            source=[self._event_source],
            detail_type=[self._event_detail_type],
        )

    def get_targets(self) -> Optional[List[IRuleTarget]]:
        return [SqsQueue(self._queue)]

File: src/stages/sqs-lambda.ts
import * as cdk from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as events from "aws-cdk-lib/aws-events";
import * as events_targets from "aws-cdk-lib/aws-events-targets";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as lambda_event_sources from "aws-cdk-lib/aws-lambda-event-sources";
import * as sqs from "aws-cdk-lib/aws-sqs";
import { Construct } from "constructs";
import { DataStage, DataStageProps } from "../pipelines/stage";

/**
 * Properties for the Lambda Function created by `SqsToLambdaStage`.
 */
export interface SqsToLambdaStageFunctionProps extends lambda.FunctionProps {
  /**
   * Amount of errored function invocations before triggering CloudWatch alarm.
   * @default 5
   */
  readonly errorsAlarmThreshold?: number;
  /**
   * The number of periods over which data is compared to the specified threshold.
   * @default 1
   */
  readonly errorsEvaluationPeriods?: number;
  /**
   * Comparison operator for evaluating alarms.
   * @default cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD
   */
  readonly errorsComparisonOperator?: cloudwatch.ComparisonOperator;
}

/**
 * Properties for `SqsToLambdaStage`.
 */
export interface SqsToLambdaStageProps extends DataStageProps {
  /**
   * Preexisting Lambda Function to use in stage.
   * If not provided, a new function will be created.
   */
  readonly lambdaFunction?: lambda.IFunction;
  /**
   * Properties for the Lambda Function that will be created by this construct
   * (if `lambdaFunction` is not provided).
   */
  readonly lambdaFunctionProps?: SqsToLambdaStageFunctionProps;
  /**
   * Preexisting SQS Queue to use in stage.
   * If not provided, a new queue will be created.
   */
  readonly sqsQueue?: sqs.IQueue;
  /**
   * Properties for the SQS Queue that will be created by this construct
   * (if `sqsQueue` is not provided).
   */
  readonly sqsQueueProps?: sqs.QueueProps;

  /**
   * The maximum number of records retrieved from the event source at the function invocation time.
   * @default 10
   */
  readonly batchSize?: number;
  /**
   * The maximum amount of time to gather records before invoking the function.
   * Valid Range: Minimum value of 0 minutes, maximum value of 5 minutes.
   * Default: - no batching window.
   */
  readonly maxBatchingWindow?: cdk.Duration;
  /**
   * Determines if DLQ is enabled.
   * @default false
   */
  readonly dlqEnabled?: boolean;
  /**
   * The number of times a message can be unsuccessfully dequeued before
   * being moved to the dead-letter queue.
   * @default 1
   */
  readonly maxReceiveCount?: number;
  /**
   * Message Group ID for messages sent to this queue.
   * Required for FIFO queues.
   */
  readonly messageGroupId?: string;
}

/**
 * Stage implements an Amazon SQS queue connected to an AWS Lambda function, with an optional DLQ.
 */
export class SqsToLambdaStage extends DataStage {
  readonly targets?: events.IRuleTarget[];
  readonly eventPattern?: events.EventPattern;

  readonly function: lambda.IFunction;
  readonly queue: sqs.IQueue;
  readonly deadLetterQueue?: sqs.Queue;

  /**
   * Constructs `SqsToLambdaStage`.
   * @param scope Scope within which this construct is defined.
   * @param id Identifier of the stage.
   * @param props Properties for the stage.
   */
  constructor(scope: Construct, id: string, props: SqsToLambdaStageProps) {
    super(scope, id, props);

    const eventSource = `${id}-event-source`;
    const eventDetailType = `${id}-event-type`;

    if (props.lambdaFunction) {
      this.function = props.lambdaFunction;
    } else if (props.lambdaFunctionProps) {
      const functionProps: SqsToLambdaStageFunctionProps = props.lambdaFunctionProps;

      this.function = new lambda.Function(this, "Process Function", {
        timeout: functionProps.timeout ?? cdk.Duration.seconds(120),
        memorySize: functionProps.memorySize ?? 256,
        environment: {
          EVENT_SOURCE: eventSource,
          EVENT_DETAIL_TYPE: eventDetailType,
          ...(functionProps.environment ?? {}),
        },
        ...functionProps,
      });
    } else {
      throw TypeError("'lambdaFunction' or 'lambdaFunctionProps' must be set to instantiate this stage");
    }

    // Enable the function to publish events to the default EventBus
    this.function.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["events:PutEvents"],
        resources: ["*"],
      }),
    );

    const dlqEnabled = props.dlqEnabled ?? false;
    if (dlqEnabled == true) {
      this.deadLetterQueue = new sqs.Queue(this, "Dead Letter Queue", {
        fifo: props.sqsQueueProps?.fifo ? props.sqsQueueProps?.fifo : undefined,
      });
    }

    if (props.sqsQueue) {
      this.queue = props.sqsQueue;
    } else {
      this.queue = new sqs.Queue(this, "Queue", {
        visibilityTimeout: props.sqsQueueProps?.visibilityTimeout ?? cdk.Duration.minutes(2),
        deadLetterQueue: this.deadLetterQueue
          ? {
              queue: this.deadLetterQueue,
              maxReceiveCount: props.maxReceiveCount ?? 1,
            }
          : undefined,
        fifo: props.sqsQueueProps?.fifo ? props.sqsQueueProps?.fifo : undefined,
        ...props.sqsQueueProps,
      });
    }

    this.function.addEventSource(
      new lambda_event_sources.SqsEventSource(this.queue, {
        batchSize: props.batchSize,
        maxBatchingWindow: props.maxBatchingWindow,
      }),
    );

    this.addAlarm("Process Function Errors", {
      metric: this.function.metricErrors(),
      threshold: props.lambdaFunctionProps?.errorsAlarmThreshold ?? 5,
      evaluationPeriods: props.lambdaFunctionProps?.errorsEvaluationPeriods ?? 1,
      comparisonOperator:
        props.lambdaFunctionProps?.errorsComparisonOperator ?? cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
    });

    this.eventPattern = {
      source: [eventSource],
      detailType: [eventDetailType],
    };

    if (this.queue.fifo && !props.messageGroupId) {
      throw TypeError("'messageGroupId' must be set to when target is a fifo queue");
    }

    this.targets = this.queue.fifo
      ? [new events_targets.SqsQueue(this.queue, { messageGroupId: props.messageGroupId })]
      : [new events_targets.SqsQueue(this.queue)];
  }
}
